consensus/polybft/consensus_runtime.go

```diff
@@ -354,6 +354,40 @@ func (c *consensusRuntime) FSM() error {
 	isEndOfSprint := c.isEndOfSprint(pendingBlockNumber)
 	isEndOfEpoch := c.isEndOfEpoch(pendingBlockNumber)
 
+	valSet, err := NewValidatorSet(epoch.Validators)
+	if err != nil {
+		return fmt.Errorf("cannot create validator set for fsm: %w", err)
+	}
+
+	if epoch.Number > 0 { // it must be at least epoc
+		iterationNumber := uint64(0)
+		currentHeader := parent
+		lastBlockOfPreviousEpoch := getEndEpochBlockNumber(epoch.Number-1, c.config.PolyBFTConfig.EpochSize)
+
+		for currentHeader.Number > lastBlockOfPreviousEpoch {
+			blockExtra, err := GetIbftExtra(currentHeader.ExtraData)
+			if err != nil {
+				return fmt.Errorf("cannot get ibft extra: %w", err)
+			}
+
+			iterationNumber += blockExtra.Round + 1 // because round 0 is one of the iterations
+
+			var found bool
+			currentHeader, found = c.config.blockchain.GetHeaderByNumber(currentHeader.Number - 1)
+
+			if !found {
+				return fmt.Errorf("cannot get header by number: %d", currentHeader.Number)
+			}
+		}
+
+		if iterationNumber > 0 {
+			err = valSet.IncrementProposerPriority(iterationNumber)
```

Comment:
Can we extract whole logic for incrementing proposer priority to separate function?

Reply:
Do not really understand, it is already extracted: IncrementProposerPriority